### DISCUSSION TITLE
Is There An Order To The Exception Handler?

### DISCUSSION THREAD

**Discussion by darwinlyang** [fastapi/fastapi#7207] [Questions] [answered]
### Example
Reference Example:
```Python
from fastapi import FastAPI


class BaseException(Exception):
    pass

class ChildException(BaseException):
    pass


app = FastAPI()


@app.exception_handler(ChildException)
async def ChildExceptionHandler(request, exc):
    return PlainTextResponse("Child exception occurred!", status_code=500)

@app.exception_handler(BaseException)
async def BaseExceptionHandler(request, exc):
    return PlainTextResponse("Base exception occurred!", status_code=500)


@app.get("/")
def read_root():
    raise ChildException()
```

My Use Case Example:
```Python
from fastapi import FastAPI

class CustomException(Exception):
    pass


app = FastAPI()


@app.exception_handler(CustomException)
async def CustomExceptionHandler(request, exc):
    return PlainTextResponse("Custom exception occurred!", status_code=500)

@app.exception_handler(Exception)
async def GenericExceptionHandler(request, exc):
    return PlainTextResponse("Generic exception occurred!", status_code=500)


@app.get("/")
def read_root():
    raise CustomException()    # Should always return "Custom exception occurred!"
```
### Description

While adding an exception handler, if I want to catch for both a child exception and a base exception, will the exception be caught based on what was added to the exception handler first? In my example above, would I always get the `"Child exception occurred!"` response whereas if I reversed the order where I added the `BaseExceptionHandler()` first would I always be getting `"Base exception occurred!"`?

My use case for this is I wanted to create custom exceptions and handle them to return custom responses but I also wanted to create a generic exception handler to catch any other exceptions from other libraries and log them as well.

I think this should be the case as that would behave similarly to a `try: ... except: ...` block where it would be caught by the first relevant exception type but I wanted to confirm that this would be the case.

**Comment by darwinlyang** [comment, marked as answer]
So after extensive testing it seems that the exception handler is smart enough to automatically go up the hierarchy for the exception until it finds a handler that suits its needs.

**Comment by tiangolo** [comment]
Thanks for coming back to close the issue @DarwinLYang! ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by thisisarko** [comment]
This is the [nifty piece of code](https://github.com/Kludex/starlette/blob/main/starlette/_exception_handler.py#L16-L20) that determines the precedence:

```python
def _lookup_exception_handler(exc_handlers: ExceptionHandlers, exc: Exception) -> ExceptionHandler | None:
    for cls in type(exc).__mro__:
        if cls in exc_handlers:
            return exc_handlers[cls]
    return None
```

It travels up the hierarchy to find a registered handler. 

If no custom handlers are provided, the [`ServerErrorMiddleware`](https://github.com/Kludex/starlette/blob/8bb71d9d51126a2177f6dd2f9c636d88e5beb410/starlette/middleware/errors.py#L170-L172) steps in to handle the base `Exception` class and returns a `PlainTextResponse` with error code `500`.
